ParProcCo/job_scheduler.py

```diff
@@ -183,163 +387,158 @@ def _create_template(self, i: int) -> drmaa2.JobTemplate:
-            job_env.update(TEST_PPC_DIR=test_ppc_dir)
-        jt.job_environment = job_env
-        return jt
 
-    def _wait_for_jobs(self, session: drmaa2.JobSession) -> None:
+        return JobSubmission(script=self.jobscript_command, job=job)
+
+    def fetch_and_update_state(self, job_id: int) -> JobResponseProperties:
+        ji = self.get_job(job_id)
+        self.update_status_infos(ji)
+        return ji
+
+    def wait_all_jobs_terminated(
+        self, job_ids: List[int], check_time: int
+    ) -> list[int]:
+        wait_until = time.time() + check_time
+        remaining_jobs = job_ids.copy()
+        sleep_time = min(60, check_time)
+        while len(remaining_jobs) > 0 and time.time() <= wait_until:
+            for job_id in list(remaining_jobs):
+                self.fetch_and_update_state(job_id)
+                if self.status_infos[job_id].current_state in STATEGROUP.ENDED:
+                    remaining_jobs.remove(job_id)
+            if len(remaining_jobs) > 0:
+                time.sleep(sleep_time)
+        return remaining_jobs
+
+    def wait_all_jobs_started(self, job_ids: List[int], check_time: int) -> list[int]:
+        wait_until = time.time() + check_time
+        remaining_jobs = job_ids.copy()
+        while len(remaining_jobs) > 0 and time.time() <= wait_until:
+            for job_id in list(remaining_jobs):
+                self.fetch_and_update_state(job_id)
+                if self.status_infos[job_id].current_state not in STATEGROUP.STARTING:
+                    remaining_jobs.remove(job_id)
+        return remaining_jobs
+
+    def _wait_for_jobs(
+        self,
+    ) -> None:
         max_time = int(round(self.timeout.total_seconds()))
         check_time = min(120, max_time)  # 2 minutes or less
         start_wait = max(check_time, 60)  # wait at least 1 minute
         try:
-            job_list = [status_info.job for status_info in self.status_infos]
+            remaining_jobs = list(self.status_infos)
             # Wait for jobs to start (timeout shouldn't include queue time)
-            job_list_str = ", ".join([str(job.id) for job in job_list])
-            logging.info(f"Waiting for jobs to start: {job_list_str}")
-            jobs_left = len(job_list)
-            while jobs_left > 0:
-                jobs_left -= len(session.wait_all_started(job_list, start_wait))
-                if jobs_left:
-                    logging.info(f"Jobs left to start: {jobs_left}")
-            logging.info(f"Jobs started, waiting for jobs: {job_list_str}")
+            while len(remaining_jobs) > 0:
+                remaining_jobs = self.wait_all_jobs_started(remaining_jobs, start_wait)
+                if len(remaining_jobs) > 0:
+                    logging.info(f"Jobs left to start: {remaining_jobs}")
             total_time = 0
-            while total_time < max_time and job_list:
-                session.wait_all_terminated(job_list, check_time)
+            begin_time = datetime.now()
```

Comment:
Why introduce this? Do it differ from `total_time`? (Also be consistent and use `time.time()`.)

Reply:
total_time includes the timeout for each wait function even if the timeout wasn't reached. I introduced the elapsed time to log the real time elapsed.